src/crustls.h

```diff
@@ -235,6 +235,55 @@ typedef struct rustls_verify_server_cert_params {
 
 typedef enum rustls_result (*rustls_verify_server_cert_callback)(rustls_verify_server_cert_user_data userdata, const struct rustls_verify_server_cert_params *params);
 
+/**
+ * Any context information the callback will receive when invoked.
+ */
+typedef void *rustls_session_store_userdata;
+
+/**
+ * Prototype of a callback that can be installed by the application at the
+ * `rustls_server_config` or `rustls_client_config`. This callback will be
+ * invoked by a TLS session when looking up the data for a TLS session id.
+ * `userdata` will be supplied as provided when registering the callback.
+ *
+ * The `buf` points to a buffer of `count` consecutive bytes where the
+ * callback is expected to copy the result to. The number of copied bytes
+ * needs to be written to `out_n`.
+ *
+ * If the value to copy is larger than `count`, the callback should never
+ * do a partial copy but instead remove the value from its store and
+ * act as if it was never found.
+ *
+ * The callback should return != 0 to indicate that a value was retrieved
+ * and written in its entirety into `buf`.
+ *
+ * When `remove_after` is != 0, the returned data needs to be removed
+ * from the store.
+ *
+ * NOTE: the passed in `key` and `buf` are only availabe during the
+ * callback invocation.
+ * NOTE: callbacks used in several sessions via a common config
+ * must be implemented thread-safe.
+ */
+typedef int (*rustls_session_store_get_callback)(rustls_session_store_userdata userdata, const struct rustls_slice_bytes *key, uint8_t *buf, size_t count, int remove_after, size_t *out_n);
```

Comment:
No big deal. Separating in and out clearly would give `userdata, key, remove_after, count, buf, out_n` but that would not be very readable either.

Made the return value a `rustls_result`, but looks weird to just convert it to a `bool` for rustls trait return value.

If `remove_after` has only a boolean value, let's not complicate things. I think we don't want `do_it` callbacks with an `rustls_operation` enum for all things now and future.

As to using `bool` in C: that touches on a sore sport, namely that Apache is still on `C89`(it's also why you do not see revolutionary C things as intermittent variable declarations in my code). 

Of course, `crustls` could declare C99 as its base.

Reply:
> Separating in and out clearly would give userdata, key, remove_after, count, buf, out_n but that would not be very readable either.

This is a specific exception to the coding conventions in the README:

> As a minor exception to the above: When an output parameter is a byte buffer
> (*uint8_t), the next parameter will always be a size_t denoting the size of
> the buffer. This is considered part of the output parameters even though it is
> not directly modified.

Reply:
> If remove_after has only a boolean value, let's not complicate things. I think we don't want do_it callbacks with an rustls_operation enum for all things now and future.

Fair enough. My reasoning for preferring enums over booleans / ints in function arguments is to make it clear at the call site what's happening. For example if I see:

`get_session(client_session, true, false, true)`

I don't know what all those fields do! But if I see:

`get_session(client_session, KEEP, ALLOW_EXPIRED, REQUIRE_SNI_MATCH)`

It's more clear. I wouldn't propose to have a single `rustls_operation` that tried to encompass all such values, but a specific enum for a specific situation.

Still, in this situation, there are no call sites in the C code. There is only one call site on the Rust side, and it is fairly close to where the variable is set, so the meaning is clear. And in the C code that defines the implementation, the variable will have a name chosen by the implementer (probably `remove_after`).

Reply:
I agree that it hinders readability on the callers side. Python's named args are really a nice way of tackling this.

We could change `remove_after: bool` to a `afterwards: rustls_post_op[KEEP, DELETE]`. The complication this brings is that the C implementer cannot  be 100% sure that these are the only `afterwards` values it will ever receive. And if it does, should it return a `NOT_IMPLEMENTED` or ignore it. That would need to be documented.

An alternative would be to adhere to the rustls pattern and introduce a 3rd callback type `take`, so the parameter is no longer needed in the `get` callback.

Reply:
I'm satisfied with the current version; plenty of other work to do. :-)